src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,226 @@
+
+                let events =
+                    page.Messages
+                    |> Seq.map StreamMessage.intoStreamEvent
+                    |> Seq.sortBy (fun x -> x.event.Index)
+                    |> Array.ofSeq
+
+                let batch =
+                    {
+                        firstPosition = events.[0].event.Index
+                        lastPosition = events.[events.Length - 1].event.Index
+                        messages = events
+                        isEnd = page.IsEnd
+                    }
+
+                logger.Debug("Submitting a batch of {batchSize} events, position {firstPosition} through {lastPosition}",
+                    batch.Length, batch.firstPosition, batch.lastPosition)
+
+                stats.UpdateBatch(batch)
+
+                let! cur, max = submitBatch (batch.lastPosition, commit batch, batch.messages)
+
+                stats.UpdateCurMax(cur, max)
+            else
+                logger.Debug("Empty page retrieved, nothing to submit")
+                stats.UpdateEmptyPage()
+        }
+
+    member this.Start (commitedPosition: Nullable<int64>) =
+        async {
+            // Start reporting stats
+            do! Async.StartChild stats.Start |> Async.Ignore
+
+            let mutable workItem = Work.TakeInitial
+
+            let sw = Stopwatch.StartNew()
+
+            let! ct = Async.CancellationToken
+            while not ct.IsCancellationRequested && not (sw.Elapsed > stopAfterInterval) do
+                try
+                    let! page =
+                        async {
+                            let! ct = Async.CancellationToken
+                            match workItem with
+                            | Work.Page page -> return page
+                            | Work.TakeInitial ->
+                                let initialPosition =
+                                    if commitedPosition.HasValue then commitedPosition.Value + 1L else 0L
+
+                                logger.Information("Starting reading stream from position {initialPosition}, maxBatchSize {maxBatchSize}", initialPosition, maxBatchSize)
+
+                                return! store.ReadAllForwards(initialPosition, maxBatchSize, true, ct) |> Async.AwaitTaskCorrect
+                            | Work.TakeNext page ->
+                                return! page.ReadNext(ct) |> Async.AwaitTaskCorrect
+                        }
+
+                    workItem <- Work.Page page
+
+                    // Process the page and submit the batch of messages to ingester.
+                    do! processPage page
+
+                    // If processPage was successful, ask for new page on the next iteration
+                    workItem <- Work.TakeNext page
+
+                    if page.IsEnd then
+                        do! Async.Sleep sleepInterval
+                with
+                | exc ->
+                    logger.Error(exc, "Exception while running StreamReader loop")
```

Comment:
this is def an Error if you are going to kill the app - are you? maybe even `Fatal` ? maybe include text indicating you're abending?

Reply:
Depends on the client - also moving that statement try with outside into SSSS.Run since I removed locking.

Reply:
I guess my point was that using Error level for something that might ultimately be handled by an outer layer might be annoying for someone else using the lib?

Reply:
But you said Fatal? :D Warn is acceptable?

Reply:
I am not telling you what way to fall on it - its unclear to me. Saying we need to pick from
1. lib is pushing it out to the caller to own what happens overall (then it should be a warning and the outher can take the real responsibility, with appropriate extra logging as to the conclusion)
2. lib is taking an opinionated stance - there's nothing we can do other than force abend (then it would be `Fatal`)